### PR TITLE
doc: Collection object and metadata spec

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Specification updates for the new collection objects and metadata! **This doesn't include any implementation support yet;** I wanted to get a gut check before moving forward with exact implementation details.

(This and the subsequent PRs will hopefully all get merged at once so there's never a mismatch between the documentation and the actual writer behavior.)

## Changes
- Updates `DATA_FORMAT.md` with the new planned metadata for collections and datasets.
- Adds new base types for shared metadata.
- Adds new dictionary types for the collection file.

*Estimated review size: medium, 20-30 minutes*

### CODE REVIEW

File: colorizer_data/types.py [hunk 1]
```diff
@@ -97,15 +100,64 @@ class FrameDimensions(TypedDict):
...
+    """Formatted datetime string. See `DATETIME_FORMAT`. """
+    revision: int = None
+    """
+    Revision number. Will be updated each time the dataset or collection
+    is rewritten. Starts at 0.
+    """
+    data_version: str = CURRENT_VERSION
+    """Version of the data writer utility scripts. Uses semantic versioning (e.g. v1.0.0)"""
+
+    def to_json(self) -> BaseMetadataJson:
```
Comment (toloudis) [inline colorizer_data/types.py:134]: there has got to be a more automatic way to do this - it's just transforming the property names

File: colorizer_data/types.py [hunk 2]
```diff
@@ -97,15 +100,64 @@ class FrameDimensions(TypedDict):
...
+    author: str
+    dateCreated: str
+    lastModified: str
+    revision: str
+    dataVersion: str
+
+
+@dataclass
+class BaseMetadata:
+    """Shared metadata between datasets and collection files."""
```
Comment (toloudis) [inline colorizer_data/types.py:117]: I suggest not sharing this at all and letting it be just duplicated because it's much easier to read and it easily allows for them to diverge.

File: colorizer_data/types.py [hunk 3]
```diff
@@ -97,15 +100,64 @@ class FrameDimensions(TypedDict):
...
+    date_created: str = None
+    """Formatted datetime string. See `DATETIME_FORMAT`. """
+    last_modified: str = None
+    """Formatted datetime string. See `DATETIME_FORMAT`. """
+    revision: int = None
+    """
+    Revision number. Will be updated each time the dataset or collection
+    is rewritten. Starts at 0.
+    """
+    data_version: str = CURRENT_VERSION
```
Comment (toloudis) [inline colorizer_data/types.py:131]: is this always incremented in lockstep or can collection version number be totally independent of dataset version number? I'm assuming we want them to be allowed to have different version numbers.

File: colorizer_data/types.py [hunk 4]
```diff
@@ -97,15 +100,64 @@ class FrameDimensions(TypedDict):
...
+class BaseMetadata:
+    """Shared metadata between datasets and collection files."""
+
+    name: str = None
+    description: str = None
+    author: str = None
+    date_created: str = None
+    """Formatted datetime string. See `DATETIME_FORMAT`. """
+    last_modified: str = None
+    """Formatted datetime string. See `DATETIME_FORMAT`. """
```
Comment (toloudis) [inline colorizer_data/types.py:125]: is the datetime format consistent with what JS code can parse?

File: colorizer_data/types.py [hunk 5]
```diff
@@ -97,15 +100,64 @@ class FrameDimensions(TypedDict):
...
+@dataclass
+class BaseMetadata:
+    """Shared metadata between datasets and collection files."""
+
+    name: str = None
+    description: str = None
+    author: str = None
+    date_created: str = None
+    """Formatted datetime string. See `DATETIME_FORMAT`. """
+    last_modified: str = None
```
Comment (toloudis) [inline colorizer_data/types.py:124]: I'm assuming that tooling will handle date_created and date_modified without users having to touch them explicitly?  (While allowing them to provide the data if they feel like it)

File: colorizer_data/types.py [hunk 6]
```diff
@@ -136,3 +188,33 @@ class DatasetManifest(TypedDict):
     bounds: str
     metadata: DatasetMetadata
     frames: List[str]
+
+
+class CollectionMetadataJson(BaseMetadataJson):
```
Comment (toloudis) [inline colorizer_data/types.py:193]: At this point it seems like we need to break up types.py and have an additional dataset.py (or dataset_types?) and collection.py (or collection_types?)

File: documentation/DATA_FORMAT.md [hunk 7]
```diff
@@ -174,6 +181,8 @@ These metadata parameters are used to configure additional features of the Timel
 
 Note that the interface will directly show the unit labels and does not scale or convert units from one type to another (for example, it will not convert 1000 µm to 1 mm). If you need to present your data with different units, create a (scaled) duplicate of the feature with a different unit label.
 
+If using the provided writer utility scripts, the `revision`, `dataVersion`, `dateCreated`, and `lastModified` fields will be automatically written and updated.
```
Comment (toloudis) [inline documentation/DATA_FORMAT.md:184]: 👍

File: documentation/DATA_FORMAT.md [hunk 8]
```diff
@@ -589,24 +598,33 @@ For example, if a dataset had the following tracks and outliers, the file might
...
-    { "name": <some_name_2>, "path": <some_path_2>},
-    ...
-]
+{
+    "datasets": [
+        { "name": <some_name_1>, "path": <some_path_1>},
+        { "name": <some_name_2>, "path": <some_path_2>},
+        ...
+    ],
+    "metadata": {
```
Comment (toloudis) [inline documentation/DATA_FORMAT.md:614]: this looks odd to me, I think I would remove the "metadata" layer and just put all those data fields at the top level.  We can discuss that if you want.

File: colorizer_data/types.py (hunk 2 above)
Comment (toloudis) [reply colorizer_data/types.py:117]: As I am thinking about this, I wonder if we could always force there to be a "collection".

File: colorizer_data/types.py (hunk 3 above)
Comment (toloudis) [reply colorizer_data/types.py:131]: Ooops I misread the comments. I always think the comment precedes the code.
I think data_version should be named writer_version (or [packagename]_version) and the above comment I posted applies to "revision".

File: colorizer_data/types.py (hunk 6 above)
Comment (ShrimpCryptid) [reply colorizer_data/types.py:193]: That makes sense to me, though it would be a breaking change for the existing API. AFAIK no one is using the colorizer-data package yet...?

File: colorizer_data/types.py (hunk 4 above)
Comment (ShrimpCryptid) [reply colorizer_data/types.py:125]: This is a good point, I'll update this to use the official JavaScript specification for simplicity

File: colorizer_data/types.py (hunk 5 above)
Comment (ShrimpCryptid) [reply colorizer_data/types.py:124]: Yup! I have a working prototype of this on a different branch.

File: colorizer_data/types.py (hunk 6 above)
Comment (ShrimpCryptid) [reply colorizer_data/types.py:193]: (I could also make `types` into a folder and nest the dataset and collection type files inside and make it into a nice tidy module)

File: colorizer_data/types.py (hunk 3 above)
Comment (ShrimpCryptid) [reply colorizer_data/types.py]: Added a new separate string parameter, `dataset_version` and `collection_version`.

File: colorizer_data/types.py (hunk 4 above)
Comment (ShrimpCryptid) [reply colorizer_data/types.py]: Learning right now that there's a whole can of worms behind timestamp formatting, but the new format should be ISO standard.

File: documentation/DATA_FORMAT.md (hunk 8 above)
Comment (ShrimpCryptid) [reply documentation/DATA_FORMAT.md:614]: I'd like to keep the metadata as its own file so I can share the logic that updates the dataset and collection metadata (updating revision number, date last modified, creation date, etc.). I imagine it'll be a little easier to reason about both that way.

File: documentation/DATA_FORMAT.md [hunk 9]
```diff
@@ -158,11 +158,16 @@ Besides the details shown above, these are additional parameters that the manife
 {
     ...
     "metadata": {
-        "frameDims": {
-            "units": <unit label for frame dimensions>,
-            "width": <width of frame in units>,
-            "height": <height of frame in units>
-        },
```
Comment (ShrimpCryptid) [inline documentation/DATA_FORMAT.md:165]: Note breaking change here, because the `dataclasses-json` library I started using doesn't have any easy way to do mapping. I have a design question about this, so let me spell out what this means:

BEFORE:
`ColorizerMetadata` accepted `frame_units`, `frame_width`, and `frame_height` arguments and wrapped them together into `frameDims` in the final output:

```
# Client API
metadata = ColorizerMetadata(frame_units = "a", frame_width = 10.0, frame_height = 20.0)

# Data API
metadata.to_json()
"""
{
    "frameDims": {
        "units": "a",
        "width": 10.0,
        "height": 20.0,
    }
}
"""
```

I can either change the client format or the data format (what I'm currently doing).

**Changing client format:**
```
metadata = ColorizerMetadata(frame_dims={"width": 10.0, "height": 20.0, "units": "a"})
```
**Changing data format:**
```
"""
{
    "frameUnits": "a",
    "frameWidth": 10.0,
    "frameHeight": 20.0,
}
"""
```

I **think** the second pattern is cleaner and I won't have to change any of the data scripts, but the tradeoff is that the colorizer UI will need to handle both the new and old data format. Any thoughts?

File: colorizer_data/types.py (hunk 6 above)
Comment (toloudis) [reply colorizer_data/types.py]: As we discussed, it probably should still be split up but can be a separate PR.  I think generally if you put imports into `__init__.py` then the things imported are considered top level exports of that directory.  So a library providing classes could import a bunch of them into its init.py file and then consumers of the package will see them as top level exports.  At least I _think_ that's how it works!  (And I don't want to worry about getting too cute with all this; just wanted to not have too many classes all in the same file.)

File: documentation/DATA_FORMAT.md (hunk 8 above)
Comment (toloudis) [reply documentation/DATA_FORMAT.md:612]: I didn't mean to change the breakdown of files. I meant that within this collection.json file I wouldn't have a "metadata" sub-object.  I'd just promote the metadata fields to be DATA FIELDS of the COLLECTION.  At the same hierarchichal level as `"datasets"`

Comment (toloudis) [reply documentation/DATA_FORMAT.md:612]: The main reason is: what is metadata?  In this sense it just feels like an artificial grouping.  Would there be any fields that are not metadata?

Comment (ShrimpCryptid) [reply documentation/DATA_FORMAT.md:612]: Oops, I mistyped! Meant to say keeping the metadata as its own field, so yes we're on the same page. 

Do you think it would be strange to have that be inconsistent between the dataset and collection files? The dataset manifests have metadata stored under the `metadata` field currently.

File: documentation/DATA_FORMAT.md (hunk 9 above)
Comment (toloudis) [reply documentation/DATA_FORMAT.md:165]: it's not clear to me why you need to change it.  Is it because `dataclasses-json` doesn't play nice with it?

File: documentation/DATA_FORMAT.md (hunk 8 above)
Comment (ShrimpCryptid) [reply documentation/DATA_FORMAT.md:612]: Posted this before I saw your updated comment! The way we have it in the dataset JSON, the root level fields are all references to files or lists of file-related descriptors; anything that's not pointing to file paths is stored under the metadata field.

I don't expect we'll add additional files to the collection schema (besides maybe a markdown description file or something in the future?), but it feels reasonable to me to be consistent.

File: documentation/DATA_FORMAT.md (hunk 9 above)
Comment (ShrimpCryptid) [reply documentation/DATA_FORMAT.md:165]: Got a working solution! (it's a little bit hacky, but there's unit tests for it.) Closing this :')

File: colorizer_data/tests/test_types.py
Comment (ShrimpCryptid) [inline colorizer_data/tests/test_types.py:1]: I have a lot more unit tests in my prototype branch that checks how `writer` interacts with the new `types`. This is a little preview specifically for the `frameDims` edge case.

File: documentation/DATA_FORMAT.md (hunk 8 above)
Comment (toloudis) [reply documentation/DATA_FORMAT.md:614]: After more discussion, it still looks odd to me, but much less odd.  I'm good to go with it!

File: documentation/DATA_FORMAT.md (hunk 9 above)
Comment (toloudis) [reply documentation/DATA_FORMAT.md]: My only preference is to not let the third party tool dictate how we want our data format to look.  And I think your working solution accomplishes that :)

File: colorizer_data/types.py [hunk 10]
```diff
@@ -139,20 +139,22 @@ class ColorizerMetadata(DataClassJsonMixin):
     author: Optional[str] = None
     dataset_version: Optional[str] = None
     """User-defined dataset version."""
-    date_created: Optional[str] = None
+
+    # Internal use
+    _date_created: Optional[str] = None
     """ISO-formatted datetime string in UTC. See `DATETIME_FORMAT`."""
-    last_modified: Optional[str] = None
+    _last_modified: Optional[str] = None
```
Comment (toloudis) [inline colorizer_data/types.py:146]: There is one interesting twist on this - the scientists might want something they can display that isn't LITERALLY the last time the file was modified but more like a "last revision date of our dataset" which could be tied to a release or paper or something.  TBH, date created and last modified are filesystem-level attributes anyway.  So maybe these could be not necessarily internal?

**Review by toloudis — APPROVED** [review]

File: colorizer_data/types.py (hunk 10 above)
Comment (ShrimpCryptid) [reply colorizer_data/types.py:146]: That makes sense to me, I can revert this change for the date-related fields

**Review by frasercl — APPROVED** [review]

File: colorizer_data/tests/test_types.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    # crash (!!!) and that values can be safely stored and retrieved.
+    frame_dimensions = [
+        {"width": 100.0, "height": 50.0, "units": "picometers"},
+        {"width": None, "height": 40.0, "units": "meters"},
+        {"height": 30.0, "units": None},
+        {},
+    ]
+
+    for i in range(len(frame_dimensions)):
+        frame_dims = frame_dimensions[i]
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply colorizer_data/tests/test_types.py:21]: Yes!

**Comment by ShrimpCryptid** [comment]
Thank you both for reviewing! Going to leave this open until the next two PRs are approved so we don't end up with a desync between the documentation and the behavior.
